tools/audio/soundfont_compiler.c

```diff
@@ -0,0 +1,1807 @@
+        if (instr_node->type != XML_ELEMENT_NODE)
+            continue;
+
+        const char *name = XMLSTR_TO_STR(instr_node->name);
+
+        bool is_instr = strequ(name, "Instrument");
+        bool is_instr_unused = strequ(name, "InstrumentUnused");
+
+        if (!is_instr && !is_instr_unused)
+            error("Unexpected element node %s in instrument list (line %d)", name, instr_node->line);
+
+        instr_data *instr = malloc(sizeof(instr_data));
+
+        instr->program_number = (unsigned)-1;
+        instr->name = NULL;
+        instr->sample_name_low = NULL;
+        instr->sample_name_mid = NULL;
+        instr->sample_name_high = NULL;
+        instr->sample_low_end = INSTR_LO_NONE;
+        instr->sample_low = NULL;
+        instr->sample_high_start = INSTR_HI_NONE;
+        instr->sample_high = NULL;
+        instr->base_note_mid = NOTE_UNSET;
+        instr->base_note_lo = NOTE_UNSET;
+        instr->base_note_hi = NOTE_UNSET;
+        instr->sample_rate_mid = -1.0;
+        instr->sample_rate_lo = -1.0;
+        instr->sample_rate_hi = -1.0;
+        instr->release = RELEASE_UNSET;
+        instr->unused = is_instr_unused;
+
+        xml_parse_node_by_spec(instr, instr_node, instr_spec, ARRAY_COUNT(instr_spec));
+
+        if (!is_instr_unused) {
+            // Check program number, midi program number range is 0-127 but the audio driver reserves 126 and 127 for
+            // sfx and percussion so the range we allow is 0-125
+            if (instr->program_number >= 126)
+                error("Program numbers must be in the range 0-125 (got %u)", instr->program_number);
+
+            // Ensure program number is unique
+            unsigned upper = instr->program_number >> 5 & 3;
+            unsigned lower = instr->program_number & 0x1F;
+            if (sf->program_number_bitset[upper] & (1 << lower))
+                error("Duplicate program number %u", instr->program_number);
+            sf->program_number_bitset[upper] |= (1 << lower);
+
+            if (instr->program_number >= sf->info.num_instruments)
+                sf->info.num_instruments = instr->program_number + 1;
+
+            // Check name
+            if (instr->name == NULL)
+                error("Instrument must be named");
+        }
+
+        // Check envelope
+        instr->envelope = sf_get_envelope(sf, instr->envelope_name);
+        if (instr->envelope == NULL)
+            error("Bad envelope name %s (line %d)\n", instr->envelope_name, instr_node->line);
+
+        // Validate optionals
+        if (instr->release == RELEASE_UNSET)
+            instr->release = instr->envelope->release;
+
+        if (instr->sample_name_mid == NULL) {
+            // For a used instrument to have no sample path, it must have sample children and have specified at least
+            // one of RangeLo or RangeHi
+
+            if (instr->sample_low_end == INSTR_LO_NONE && instr->sample_high_start == INSTR_HI_NONE)
+                error("epic fail");
```

Comment:
😐

Reply:
Improved a lot of the error messages just for you

Reply:
thank you friendddd